### ISSUE TITLE
Can't run ava on cli with a glob in v2

### ISSUE THREAD

**Issue by StarpTech** [avajs/ava#2164]
Hello,
I just wanted to upgrade to v2 but I got an error when trying to run

```
**/*.spec.js => unit tests
**/*.spec.int.js => for integration tests running only in CI or on demand
```
in v1.4.1 it works!

Error:
```
  ✖ Internal error
  Error: ENOENT: no such file or directory, lstat '/mnt/ssd/repositories/test/lca/packages/lca/**'
  Error: ENOENT: no such file or directory, lstat '/mnt/ssd/repositories/test/lca/packages/lca/**'
      at Object.realpathSync (fs.js:1470:7)
      at /mnt/ssd/repositories/test/lca/node_modules/ava/lib/api.js:186:28
      at Array.reduce (<anonymous>)
      at Api.run (/mnt/ssd/repositories/test/lca/node_modules/ava/lib/api.js:184:34)
      at processTicksAndRejections (internal/process/task_queues.js:89:5)
      at async Object.exports.run (/mnt/ssd/repositories/test/lca/node_modules/ava/lib/cli.js:274:21)
```

## Environment

```
Node.js v12.4.0
linux 4.18.0-22-generic
ava: 2.1.0
npm: 6.9.0
```

**Comment by sindresorhus** [comment]
> The CLI now only takes file paths, not glob patterns. - https://github.com/avajs/ava/releases/tag/v2.0.0

**Comment by StarpTech** [comment]
@sindresorhus so how can I differentiate between different test types in my project? e.g

```
*.spec.js => unit tests
*.spec.int.js => for integration tests running only in CI or on demand
```

**Comment by StarpTech** [comment]
Related

* https://github.com/avajs/ava/issues/2152
* https://github.com/avajs/ava/issues/2162

**Comment by novemberborn** [comment]
Hi @StarpTech, as you probably found when reading https://github.com/avajs/ava/issues/2152 I hadn't considered that use case when removing glob support from the CLI. I've now opened a PR that supports a `--config` flag on the CLI, allowing you to provide different config depending on whether you're running unit or integration tests: https://github.com/avajs/ava/pull/2166

Perhaps you could give it a try to see if it solves your use case?

(I'm closing this issue for housekeeping purposes, but let's keep the conversation going.)

**Comment by StarpTech** [comment]
Hi @novemberborn thanks for the info. What was the reason to remove it? In my opinion it is very important for UX and the defacto standard how test runners are used. The config flag will solve it but it adds some extra bloat.

**Comment by novemberborn** [comment]
> What was the reason to remove it?

I wanted to simplify how we find test and helper files, making it easier to understand for users and maintainers alike.

> In my opinion it is very important for UX and the defacto standard how test runners are used.

FWIW you can still use your shell's globbing. There's some portability issues with that which depend on what operating systems and shells your team uses.

**Comment by gauntface** [comment]
This is a real shame :(

If anyone can provide an example for how to use the config flag i'd appreciate it.

**Comment by novemberborn** [comment]
> If anyone can provide an example for how to use the config flag i'd appreciate it.

Hi @gauntface, the flag is not out yet, but hopefully soon. For my understanding, how were you using this with AVA 1.0?

**Comment by bookmoons** [comment]
Started upgrading but this breaks my testing pattern. Sticking with v1 for now. I'm really depending on being able to specify tests ad hoc while working on a certain component.

```json
{
    "scripts": {
        "test-hoc": "ava"
    }
}
```

```sh
npm run test-hoc test/unit/client
npm run test-hoc test/unit/server
npm run test-hoc test/unit/server/listener/tcp
npm run test-hoc test/int/reader
```

Etc. I switch the tests I'm running several times every day. Specify a directory and I get everything under that dir. It's a really good UX.

**Comment by novemberborn** [comment]
> Etc. I switch the tests I'm running several times every day. Specify a directory and I get everything under that dir. It's a really good UX.

Since this is a local use case, you should be able to use globbing in your shell:

```console
npm run test-hoc test/unit/client/**/*
```

**Comment by bookmoons** [comment]
That doesn't seem to catch tests at the top level.

```
test
  unit
    server
      start.js
      stop.js
      initiator
        standard.js
```

```sh
$ ls test/unit/server/**/*
test/unit/server/initiator/standard.js
```

It only selects tests in subdirs.

**Comment by novemberborn** [comment]
This might work: `test/unit/client/{*,**/*}` but maybe you need to list the extension, so `test/unit/client/{*.js,**/*.js}`.

**Comment by grnch** [comment]
Hello @novemberborn,

Are you seriously suggesting that people type those horrible command lines every time they want to run Ava on a subset of their test suite??

Not to mention that this leaves people on Windows completely SOL.

I just upgraded a project to Ava 2.x and got this really nasty surprise... :flushed:

My use case is very similar to @bookmoons above. We have a rather large test suite that is split up in various folders (and subfolders) depending on which area of code is being tested.

It's a very common workflow to be able to run a specific folder or subfolder of tests for the specific area of code you might be currently working on.

The `--config` flag is fairly useless in this case, because we're not talking about permanently fixed subsets of tests, where you can create the config files in advance. The subset of tests you might need to run changes from day to day and even hour to hour.

I don't understand why the simple ability to specify a folder on the CLI was removed?

You mention above that the reason for the removal was "making it easier to understand for users and maintainers alike", but how can this atrocity be simpler to understand and remember for users?

 `ava test/unit/client/{*,**/*}`

Compared to something as simple and clear as this?

 `ava test/unit/client`

By any reasonable UX measure this is a change for the worse. More difficult to read, more difficult to type, more difficult to understand, and more difficult to remember.

Please, please, please consider bringing the old CLI behavior back.

🙏 🙏 🙏

**Comment by novemberborn** [comment]
What we could do is keep relying on the configured glob patterns, *but* let you specify a subset on the CLI, and when we select the actual tests we can support file and folder matches (maybe even with some pattern support). What do you think? Perhaps you could open a new issue for that solution?

**Comment by grnch** [comment]
Filtering the configured glob patterns by folders (or glob patterns) specified on the CLI would certainly improve quality of life, especially for Windows developers. Thank you for being willing to take a second look at this!

My only concern would be that it would make this interaction between configured globs and CLI parameters more difficult to document and explain.

All the other CLI parameters work pretty much exactly the same as their config equivalents (e.g. the `--concurrency` CLI argument works exactly the same as the `concurrency` option specified in the JSON config).

But in this case, the `files` option in the JSON config would have a different meaning than the list of files specified on the CLI.

However, if this discrepancy doesn't bother you, then your proposal would certainly solve the immediate problem of not being able to run arbitrary subfolders of a test suite. I'm all for it! 👍

**Comment by novemberborn** [comment]
> But in this case, the `files` option in the JSON config would have a different meaning than the list of files specified on the CLI.

Sure, but the way the files are interpreted also depends on configured extensions and whatnot. That's one of the reasons globs were removed from the CLI! The tricky bit is finding a definition of the CLI behavior that works with the configured behavior. I think we've hit upon it with https://github.com/avajs/ava/issues/2164#issuecomment-545806434.

I've opened https://github.com/avajs/ava/issues/2281 to track.
